cpp/include/cudf/detail/indexalator.cuh

```diff
@@ -0,0 +1,454 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/utilities/traits.hpp>
+
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief The base class for the input or output index normalizing iterator.
+ *
+ * This implementation uses CTRP to define the input_indexalator and the
+ * output_indexalator classes. This is so this class can manipulate the
+ * uniquely typed subclass member variable `p_` directly without requiring
+ * virtual functions since iterator instances will be copied to device memory.
+ *
+ * The base class mainly manages updating the `p_` member variable while the
+ * subclasses handle accessing individual elements in device memory.
+ *
+ * @tparam T The derived class type for the iterator.
+ */
+template <class T>
+struct base_indexalator {
```

Comment:
I think you may be able to eliminate a lot of this code using thrust's `iterator_adapter` as a base class.

https://github.com/thrust/thrust/blob/main/thrust/iterator/iterator_adaptor.h

Reply:
Also, `transform_input_output_iterator` was added to Thrust 1.10.0 RC. Wondering if that would be useful here (in the future).

https://github.com/thrust/thrust/blob/main/thrust/iterator/transform_input_output_iterator.h

Reply:
The `iterator_adapter` cannot be used since it defeats the purpose of providing a single class instance for any integer type. If we were to try to use it for our fixed `size_type` it would not provide the proper logic for the operators.

Reply:
There's a very good possibility I don't understand the use case... but just to be clear, I was suggesting deriving it and overriding as much functionality as necessary. There's a lot of boilerplate here that maybe we can get rid of? Again, I might just misunderstand the use case. Just checking.